File: MapsSDKExamplesObjc/Driving/ZoomingTheRouteViewController.m
/**
 * Copyright (c) 2021 TomTom N.V. All rights reserved.
 *
 * This software is the proprietary copyright of TomTom N.V. and its
 * subsidiaries and may be used for internal evaluation purposes or commercial
 * use strictly subject to separate licensee agreement between you and TomTom.
 * If you are the licensee, you are only permitted to use this Software in
 * accordance with the terms of your license agreement. If you are not the
 * licensee then you are not authorised to use this software in any manner and
 * should immediately return it to TomTom N.V.
 */

#import "ZoomingTheRouteViewController.h"
#import <TomTomOnlineSDKMaps/TomTomOnlineSDKMaps.h>

#define ZOOM_CHANGE_ANIMATION_MILLIS 300
#define SMALL_SPEED_ZOOM_LEVEL 19.0
#define MEDIUM_SPEED_ZOOM_LEVEL 18.0
#define GREATER_SPEED_ZOOM_LEVEL 17.0
#define BIG_SPEED_ZOOM_LEVEL 16.0
#define HUGE_SPEED_ZOOM_LEVEL 14.0
#define SMALL_SPEED_RANGE_END_IN_KMH 10.0
#define MEDIUM_SPEED_RANGE_END_IN_KMH 20.0
#define GREATER_SPEED_RANGE_END_IN_KMH 40.0
#define BIG_SPEED_RANGE_END_IN_KMH 70.0
#define HUGE_SPEED_RANGE_END_IN_KMH 120.0

@interface ZoomingTheRouteViewController () <TTRouteResponseDelegate, LocationManagerMockDelegate>

@property(nonatomic, strong) TTChevronObject *chevron;
@property(nonatomic, strong) TTRoute *routePlanner;

@property(nonatomic, strong) LocationManagerMock *source;
@property(nonatomic, assign) BOOL updateZoom;
@property(nonatomic) CLLocationCoordinate2D *waypoints;

@end

@implementation ZoomingTheRouteViewController

- (void)setupInitialCameraPosition {
    [self.mapView centerOnCoordinate:TTCoordinate.LODZ withZoom:10];
}

- (OptionsView *)getOptionsView {
    return [[OptionsViewSingleSelect alloc] initWithLabels:@[ @"Start tracking", @"Stop tracking" ] selectedID:-1];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    _routePlanner = [[TTRoute alloc] initWithKey:Key.Routing];
    self.waypoints = malloc(sizeof(CLLocationCoordinate2D) * 3);
    self.waypoints[0] = [TTCoordinate LODZ_SREBRZYNSKA_WAYPOINT_A];
    self.waypoints[1] = [TTCoordinate LODZ_SREBRZYNSKA_WAYPOINT_B];
    self.waypoints[2] = [TTCoordinate LODZ_SREBRZYNSKA_WAYPOINT_C];
    self.routePlanner.delegate = self;
    self.mapView.contentInset = TTCamera.MAP_DEFAULT_INSETS;
    [self.progress show];
}

- (void)viewWillDisappear:(BOOL)animated {
    [self.source deactivate];
    free(self.waypoints);
    [super viewWillDisappear:animated];
}

- (void)onMapReady {
    [super onMapReady];
    [self createRoute];
}

- (void)displayExampleWithID:(NSInteger)ID on:(BOOL)on {
    [super displayExampleWithID:ID on:on];
    switch (ID) {
    case 1:
        [self stop];
        break;
    default:
        [self start];
        break;
    }
}

- (void)createRoute {
    TTRouteQuery *query = [[[TTRouteQueryBuilder createWithDest:TTCoordinate.LODZ_SREBRZYNSKA_STOP andOrig:TTCoordinate.LODZ_SREBRZYNSKA_START] withWayPoints:self.waypoints count:3] build];
    [self.routePlanner planRouteWithQuery:query];
}

- (void)createChevron {
    [self.mapView setShowsUserLocation:false];
    TTChevronAnimationOptions *animation = [[TTChevronAnimationOptionsBuilder createWithAnimatedCornerRounding:true] build];
    self.chevron = [[TTChevronObject alloc] initWithNormalImage:[TTChevronObject defaultNormalImage] withDimmedImage:[TTChevronObject defaultDimmedImage] withChevronAnimationOptions:animation];
}

- (void)start {
    self.updateZoom = YES;
    self.chevron.hidden = NO;
    TTCameraPosition *camera = [[[[[[TTCameraPositionBuilder createWithCameraPosition:[TTCoordinate LODZ_SREBRZYNSKA_START]] withAnimationDuration:[TTCamera ANIMATION_TIME]] withBearing:[TTCamera BEARING_START]] withPitch:[TTCamera DEFAULT_MAP_PITCH_LEVEL_FOR_DRIVING]]
        withZoom:[TTCamera DEFAULT_MAP_ZOOM_LEVEL_FOR_DRIVING]] build];

    [self.mapView setCameraPosition:camera];

    if (!self.chevron) {
        return;
    }
    [self.mapView.trackingManager setBearingSmoothingFilter:[TTTrackingManagerDefault bearingSmoothFactor]];
    [self.mapView.trackingManager startTrackingObject:self.chevron];
}

- (void)stop {
    self.updateZoom = NO;
    [self.mapView.trackingManager stopTrackingObject:self.chevron];
    [self displayRouteOverview];
}

#pragma mark TTRouteResponseDelegate

- (void)route:(TTRoute *)route completedWithResult:(TTRouteResult *)result {
    TTFullRoute *plannedRoute = result.routes.firstObject;
    if (!plannedRoute) {
        return;
    }
    TTMapRoute *mapRoute = [TTMapRoute routeWithCoordinatesData:result.routes.firstObject withRouteStyle:TTMapRouteStyle.defaultActiveStyle imageStart:TTMapRoute.defaultImageDeparture imageEnd:TTMapRoute.defaultImageDestination];
    [self.mapView.routeManager addRoute:mapRoute];
    [self.mapView.routeManager bringToFrontRoute:mapRoute];
    [self.etaView showWithSummary:plannedRoute.summary style:ETAViewStylePlain];
    [self displayRouteOverview];
    [self.progress hide];

    [self createChevron];
    [self.mapView.trackingManager addTrackingObject:self.chevron];
    self.source = [[LocationManagerMock alloc] initWithRoute:mapRoute];
    self.source.delegate = self;
    [self.source activate];
}

- (void)route:(TTRoute *)route completedWithResponseError:(TTResponseError *)responseError {
    [self handleError:responseError];
}

- (void)applyLocation:(CLLocation *)location {
    [self.mapView.trackingManager updateTrackingObject:self.chevron withTTLocation:[[TTLocation alloc] initWithCoordinate:location.coordinate withRadius:location.horizontalAccuracy withBearing:location.course withAccuracy:location.horizontalAccuracy]];
    [self updateZoomLevelBaseOnNewLocation:location];
}

- (void)updateZoomLevelBaseOnNewLocation:(CLLocation *)location {
    if (!_updateZoom) {
        return;
    }
    double zoom;
    if (location.speed < SMALL_SPEED_RANGE_END_IN_KMH) {
        zoom = SMALL_SPEED_ZOOM_LEVEL;
    } else if (location.speed < MEDIUM_SPEED_RANGE_END_IN_KMH) {
        zoom = MEDIUM_SPEED_ZOOM_LEVEL;
    } else if (location.speed < GREATER_SPEED_RANGE_END_IN_KMH) {
        zoom = GREATER_SPEED_ZOOM_LEVEL;
    } else if (location.speed < BIG_SPEED_RANGE_END_IN_KMH) {
        zoom = BIG_SPEED_ZOOM_LEVEL;
    } else {
        zoom = HUGE_SPEED_ZOOM_LEVEL;
    }
    TTCameraPosition *cameraPosition = [[[[TTCameraPositionBuilder createWithCameraPosition:self.mapView.cameraPosition.cameraPosition] withZoom:zoom] withAnimationDuration:ZOOM_CHANGE_ANIMATION_MILLIS] build];
    [self.mapView setCameraPosition:cameraPosition];
}

- (void)didUpdateLocation:(CLLocation *)location {
    [self applyLocation:location];
}
@end

File: MapsSDKExamplesSwift/Driving/ZoomingTheRouteViewController.swift
/**
 * Copyright (c) 2021 TomTom N.V. All rights reserved.
 *
 * This software is the proprietary copyright of TomTom N.V. and its subsidiaries and may be used
 * for internal evaluation purposes or commercial use strictly subject to separate licensee
 * agreement between you and TomTom. If you are the licensee, you are only permitted to use
 * this Software in accordance with the terms of your license agreement. If you are not the
 * licensee then you are not authorised to use this software in any manner and should
 * immediately return it to TomTom N.V.
 */

import MapsSDKExamplesCommon
import MapsSDKExamplesVC
import TomTomOnlineSDKMaps
import UIKit

class ZoomingTheRouteViewController: RoutingBaseViewController, TTRouteResponseDelegate {
    let ZOOM_CHANGE_ANIMATION_MILLIS = 300
    let SMALL_SPEED_ZOOM_LEVEL = 19.0
    let MEDIUM_SPEED_ZOOM_LEVEL = 18.0
    let GREATER_SPEED_ZOOM_LEVEL = 17.0
    let BIG_SPEED_ZOOM_LEVEL = 16.0
    let HUGE_SPEED_ZOOM_LEVEL = 14.0
    let SMALL_SPEED_RANGE_IN_KMH = 0.0 ..< 10.0
    let MEDIUM_SPEED_RANGE_IN_KMH = 10.0 ..< 20.0
    let GREATER_SPEED_RANGE_IN_KMH = 20.0 ..< 40.0
    let BIG_SPEED_RANGE_IN_KMH = 40.0 ..< 70.0
    let HUGE_SPEED_RANGE_IN_KMH = 70.0 ..< 120.0
    let routePlanner = TTRoute(key: Key.Routing)
    var waypoints = [TTCoordinate.LODZ_SREBRZYNSKA_WAYPOINT_A(),
                     TTCoordinate.LODZ_SREBRZYNSKA_WAYPOINT_B(),
                     TTCoordinate.LODZ_SREBRZYNSKA_WAYPOINT_C()]

    var source: LocationManagerMock?

    var updateZoom = false
    var chevron: TTChevronObject?

    override func setupInitialCameraPosition() {
        mapView.center(on: TTCoordinate.LODZ(), withZoom: 10)
    }

    override func getOptionsView() -> OptionsView {
        return OptionsViewSingleSelect(labels: ["Start tracking", "Stop tracking"], selectedID: -1)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        routePlanner.delegate = self
        mapView.contentInset = TTCamera.MAP_DEFAULT_INSETS()
        progress.show()
    }

    override func viewWillDisappear(_ animated: Bool) {
        source?.deactivate()
        super.viewWillDisappear(animated)
    }

    override func onMapReady() {
        super.onMapReady()
        createRoute()
    }

    override func displayExample(withID ID: Int, on: Bool) {
        super.displayExample(withID: ID, on: on)
        switch ID {
        case 1:
            stop()
        default:
            start()
        }
    }

    func createRoute() {
        let query = TTRouteQueryBuilder.create(withDest: TTCoordinate.LODZ_SREBRZYNSKA_STOP(), andOrig: TTCoordinate.LODZ_SREBRZYNSKA_START())
            .withWayPoints(&waypoints, count: UInt(waypoints.count))
            .build()
        routePlanner.plan(with: query)
    }

    func createChevron() {
        self.mapView.isShowsUserLocation = false
        let animation = TTChevronAnimationOptionsBuilder.create(withAnimatedCornerRounding: false).build()
        chevron = TTChevronObject(normalImage: TTChevronObject.defaultNormalImage(), withDimmedImage: TTChevronObject.defaultDimmedImage(), with: animation)
    }

    func start() {
        updateZoom = true
        chevron!.isHidden = false
        let camera = TTCameraPositionBuilder.create(withCameraPosition: TTCoordinate.LODZ_SREBRZYNSKA_START())
            .withAnimationDuration(TTCamera.ANIMATION_TIME())
            .withBearing(TTCamera.BEARING_START())
            .withPitch(TTCamera.DEFAULT_MAP_PITCH_LEVEL_FOR_DRIVING())
            .withZoom(Double(TTCamera.DEFAULT_MAP_ZOOM_LEVEL_FOR_DRIVING()))
            .build()
        mapView.setCameraPosition(camera)

        guard let chevron = self.chevron else { return }
        mapView.trackingManager.setBearingSmoothingFilter(TTTrackingManagerDefault.bearingSmoothFactor())
        mapView.trackingManager.start(chevron)
    }

    func stop() {
        updateZoom = false
        mapView.trackingManager.stop(chevron!)
        displayRouteOverview()
    }

    // MARK: TTRouteResponseDelegate

    func route(_: TTRoute, completedWith result: TTRouteResult) {
        guard let plannedRoute = result.routes.first else {
            return
        }
        let mapRoute = TTMapRoute(coordinatesData: plannedRoute,
                                  with: TTMapRouteStyle.defaultActive(),
                                  imageStart: TTMapRoute.defaultImageDeparture(),
                                  imageEnd: TTMapRoute.defaultImageDestination())

        mapView.routeManager.add(mapRoute)
        mapView.routeManager.bring(toFrontRoute: mapRoute)
        displayRouteOverview()
        progress.hide()

        createChevron()
        mapView.trackingManager.add(self.chevron!)
        source = LocationManagerMock(route: mapRoute)
        source?.delegate = self
        source?.activate()
    }

    func route(_: TTRoute, completedWith responseError: TTResponseError) {
        handleError(responseError)
    }

    func applyLocation(_ location: CLLocation) {
        mapView.trackingManager.update(chevron!, with: TTLocation(coordinate: location.coordinate, withRadius: location.horizontalAccuracy, withBearing: location.course, withAccuracy: location.horizontalAccuracy))
        updateZoomLevelBaseOnNewLocation(location)
    }

    func updateZoomLevelBaseOnNewLocation(_ location: CLLocation) {
        guard updateZoom else { return }
        var zoom: Double
        switch location.speed {
        case SMALL_SPEED_RANGE_IN_KMH:
            zoom = SMALL_SPEED_ZOOM_LEVEL
        case MEDIUM_SPEED_RANGE_IN_KMH:
            zoom = MEDIUM_SPEED_ZOOM_LEVEL
        case GREATER_SPEED_RANGE_IN_KMH:
            zoom = GREATER_SPEED_ZOOM_LEVEL
        case BIG_SPEED_RANGE_IN_KMH:
            zoom = BIG_SPEED_ZOOM_LEVEL
        default:
            zoom = HUGE_SPEED_ZOOM_LEVEL
        }
        guard mapView.zoom() != zoom else { return }
        let cameraPostion = TTCameraPositionBuilder.create(withCameraPosition: mapView.cameraPosition().cameraPosition).withZoom(zoom).withAnimationDuration(Int32(ZOOM_CHANGE_ANIMATION_MILLIS)).build()
        mapView.setCameraPosition(cameraPostion)
    }

}

extension ZoomingTheRouteViewController: LocationManagerMockDelegate {
    func didUpdateLocation(_ location: CLLocation) {
        applyLocation(location)
    }
}
